### PR TITLE
feat(application-system): add translation service

### PR DESCRIPTION
**Pull request by jeremybarbet**
### Usage

Since theses errors messages are coming from the API on pages like minarsidur or the list on umsoknir, we need to pass the locale as a context of the graphql query and refetch when the locale changes.

```typescript
const { lang: locale } = useLocale();

const { refetch } = useQuery(APPLICATION_APPLICATIONS, { variables: { input: { ... }, locale } });

useEffect(refetch, [locale]);
```

### `TranslationsService`

- Refactor to cache translations message with memoizee + return an object containing all languages

### `IntlService`

- New IntlService that get the messages fetched by the `translationsService`
- `createIntl` from `formatjs`
- Exposes the intl object to contains all the methods available
- Overwrite formatMessage for our use case.

### `applicationTemplate.ts`

- Now have a `namespaceTranslations: string[]` field to pass the custom translation to load from Contentful.
- Inside the application core we are also loading `application.system` used for generic error messages needed.
- Each state of the template now has two new fields `title` and `description` that takes a `StaticText` (string or translatable message). The active state of the application will return the current "description" if found and will be used to show on the application list on umsoknir and minarsidur (See screenshots below)

### `dataSchema.ts`

- dataSchema.ts now support translation with the following syntax thanks to zod custom error message object
```typescript
z.string().refine(v => v === 'hello', { params: namespaceName.errorMessage })
```
You can either use a generic error message coming from `coreErrorMessages` or your own from your own messages file.

### `answerValidators.ts`

- answerValidators now support translation by either importing `coreErrorMessages` or your messages form your own file
```typescript
buildValidationError('field.path')(coreErrorMessages.defaultError)
```

## Screenshots

**Mínar siður (en/is)**

Screenshot 2021-04-09 at 14 46 00
Screenshot 2021-04-09 at 14 45 52

### CODE REVIEW

File: libs/service-portal/applications/src/screens/ApplicationList/ApplicationList.tsx
Comment (jeremybarbet) [reply libs/service-portal/applications/src/screens/ApplicationList/ApplicationList.tsx:92]: I believe the designers want to have some generic messages for every application and not the custom for each application so the user can have common knowledge when they consult its list of applications

File: libs/application/core/src/types/StateMachine.ts
Comment (jeremybarbet) [reply libs/application/core/src/types/StateMachine.ts:66]: I can just go for required for both fields, so we are just to have useful data in any state of the application and to show on the service portal list

File: libs/application/core/.eslintrc
Comment (jeremybarbet) [reply libs/application/core/.eslintrc:4]: In `nx.json` along with the base `eslintrc.json`, like [this](https://nx.dev/latest/angular/structure/monorepo-tags#tags)?

File: libs/service-portal/applications/src/screens/ApplicationList/ApplicationList.tsx
Comment (eirikurn) [reply libs/service-portal/applications/src/screens/ApplicationList/ApplicationList.tsx:92]: Makes sense. Though I think it should be very obvious that the application is "Waiting for employer approval".

Maybe we can show that as a point on the progress instead?

File: libs/application/core/src/types/StateMachine.ts
Comment (eirikurn) [reply libs/application/core/src/types/StateMachine.ts:66]: Yeah. Let's do this in a follow up. Ideally most states are extending builtin states:

```
meta: {
  ...defaultStates.draft, // get title, description, lifetime and more for free
  // override stuff
}
```

File: libs/service-portal/applications/src/screens/ApplicationList/ApplicationList.tsx
Comment (jeremybarbet) [reply libs/service-portal/applications/src/screens/ApplicationList/ApplicationList.tsx]: I discussed about it with Jóhanna and she doesn't want to create confusion by showing another status than the generic ones ("completed", "rejected", and "in progress").

However, since we are likely going to show a detailed page for each application showing the different steps the application contains and the status of each of them, the `stateTitle` will be of good use there

**Review by stjanilofts — APPROVED** [review]
Kosmos code owners stuff LGTM

File: libs/api/domains/translations/src/lib/translations.service.ts
Comment (eirikurn) [inline libs/api/domains/translations/src/lib/translations.service.ts:70]: Is this check why it was using `mergeWith` before? 

I'd like to optimize this function quite a bit, to not process unused locales, and not create Namespaces * Locales * Keys growing objects.

I'd suggest moving the fallback logic behind the memoized function, so it returns an object that looks like this: `{ is: allIcelandicTranslations, en: allEnglishTranslationsWithFallbacks }`. Then we can easily loop through the requested namespaces and merge the requested locale into a final object.

I generally avoid doing `{ ...obj, [key]: value }` many times in the hot path (code that runs a lot) and modify a mutable object instead. But this is less critical if we move it behind memoizee.

File: apps/application-system/form/src/routes/Applications.tsx
Comment (jeremybarbet) [reply apps/application-system/form/src/routes/Applications.tsx]: It does work the first time you switch language, but it doesn't work when trying to switch again afterward

File: libs/api/domains/translations/src/lib/translations.service.ts
Comment (jeremybarbet) [reply libs/api/domains/translations/src/lib/translations.service.ts:45]: Good catch

File: libs/api/domains/application/src/lib/application.resolver.ts
Comment (jeremybarbet) [reply libs/api/domains/application/src/lib/application.resolver.ts:37]: That's sounds good to me 👍

File: apps/application-system/form/src/routes/Applications.tsx
Comment (jeremybarbet) [reply apps/application-system/form/src/routes/Applications.tsx]: Doesn't really change the initial problem, but I created a hook that will handle the locale ref and refetch

**Review by Thorrik58 — APPROVED** [review]
LGTM

**Review by dd-sdang — APPROVED** [review]
Mostly reviewed Health Insurance and it looks good :)

**Review by fjandakornid — APPROVED** [review]
LGTM

**Review by ingolfurs — APPROVED** [review]

File: libs/api/domains/translations/src/lib/translations.service.ts
Comment (eirikurn) [inline libs/api/domains/translations/src/lib/translations.service.ts:62]: Let's move this one level higher (witho `?? {}`). We can skip the next `if` condition as well.

File: libs/api/domains/translations/src/lib/translations.service.ts
Comment (eirikurn) [inline libs/api/domains/translations/src/lib/translations.service.ts:71]: Can move the `locale` variable one level up, as well as caching the locale strings:

```
const localeStrings = strings[contentfulLocale] ?? {}
```

Can also cache `defaultStrings` one level higher.

File: libs/api/domains/translations/src/lib/translations.service.ts
Comment (eirikurn) [inline libs/api/domains/translations/src/lib/translations.service.ts:79]: I think there might be a bug here. What if some key is undefined in English for some reason, then it will end up undefined for that locale instead of using the default locale. Should be fine to check if the string is "trueish", or just `localeStrings[key] || defaultStrings[key]`.

**Review by eirikurn — COMMENTED** [review]
Almost there! 😅 Just one potential bug, and some minor suggestions since you're touching the code anyways. :)

File: libs/api/domains/translations/src/lib/translations.service.ts
Comment (jeremybarbet) [reply libs/api/domains/translations/src/lib/translations.service.ts:79]: I'm pretty sure that's not possible. When running the extract strings script, we create strings for both locales and default to an empty string for English messages. If it has not been edited inside contentful, we will get an empty string as value, if we don't have an Icelandic locale for it neither.

But I'm only "pretty sure" so I can add the test case just to be "100% pretty sure" :D

Comment (eirikurn) [reply libs/api/domains/translations/src/lib/translations.service.ts:79]: Ahhh, makes sense. This should at least save us some grief if we add more locales later.

**Review by eirikurn — APPROVED** [review]
Sweet! Let's get this merged.

**Pull request merged**
